calicoctl/commands/node/diags.go

```diff
@@ -0,0 +1,161 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package node
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/docopt/docopt-go"
+)
+
+// diagCmds is a slice of struct to hold commands, their info and filename to run diagnostic on
+type diagCmds struct {
+	info     string
+	cmd      string
+	filename string
+}
+
+// Diags function collects diagnostic information and logs
+func Diags(args []string) error {
+	var err error
+	doc := `Usage: node diags <LOG_DIR>
+
+    Description:
+         Save diagnostic information`
+
+	arguments, err := docopt.Parse(doc, args, true, "calicoctl", false, false)
+	if err != nil {
+		return err
+	}
+
+	runDiags(arguments["<LOG_DIR>"].(string))
+
+	return nil
+}
+
+// runDiags takes logDir and runs a sequence of commands to collect diagnostics
+func runDiags(logDir string) {
+
+	// Note: in for the cmd field in this struct, it  can't handle args quoted with space in it
+	// For example, you can't add cmd "do this", since after the `strings.Fields` it will become `"do` and `this"`
+	cmds := []diagCmds{
+		{"", "date", "date"},
+		{"", "hostname", "hostname"},
+		{"Dumping netstat", "netstat --all --numeric", "netstat"},
+		{"Dumping routes (IPv4)", "ip -4 route", "ipv4_route"},
+		{"Dumping routes (IPv6)", "ip -6 route", "ipv6_route"},
+		{"Dumping interface info (IPv4)", "ip -4 addr", "ipv4_addr"},
+		{"Dumping interface info (IPv6)", "ip -6 addr", "ipv6_addr"},
+		{"Dumping iptables (IPv4)", "iptables-save", "ipv4_tables"},
+		{"Dumping iptables (IPv6)", "ip6tables-save", "ipv6_tables"},
+		{"Dumping ipsets", "ipset list", "ipsets"},
+		{"Dumping ipsets (container)", "docker run --privileged --net=host calico/node ipset list", "ipset_container"},
+		{"Copying journal for calico-node.service", "journalctl -u calico-node.service --no-pager", "journalctl_calico_node"},
+	}
+
+	// Make sure the command is run with super user priviladges
+	if os.Getuid() != 0 {
+		fmt.Println("Need super user privilages: Operation not permitted")
+		os.Exit(1)
+	}
+
+	fmt.Println("Collecting diagnostics")
+
+	// Create a temp directory in /tmp
+	tmpDir, err := ioutil.TempDir("/tmp", "tmp")
+	if err != nil {
+		log.Fatalf("Error creating temp directory to dump logs: %v\n", err)
+	}
+
+	fmt.Println("Using temp dir:", tmpDir)
+	err = os.Chdir(tmpDir)
+	if err != nil {
+		log.Fatalf("Error changing directory to temp directory to dump logs: %v\n", err)
+	}
+
+	os.Mkdir("diagnostics", os.ModeDir)
+	diagsTmpDir := filepath.Join(tmpDir, "diagnostics")
+
+	for _, v := range cmds {
+		writeDiags(v, diagsTmpDir)
+	}
+
+	fmt.Println("Dumping felix stats")
+	err = exec.Command("pkill", "-SIGUSR1", "felix").Run()
```

Comment:
yup, good point! I originally had it outside because we're not saving it's output/it doesn't produce any output to stdout, but I added it to the cmd list now